pypsa/definitions/components.py

```diff
@@ -1,120 +1,126 @@
+
+    def __getattr__(self, item: str) -> Any:
+        """
+        Get attribute from the dictionary.
+
+        Examples
+        --------
+        >>> components = ComponentsStore()
+        >>> components["generator"] = Generators()
+        >>> components.generators
+        """
+        try:
+            return self[item]
+        except KeyError:
+            msg = f"'{self.__class__.__name__}' object has no component '{item}'."
+            raise AttributeError(msg)
+
+    def __getitem__(self, item: str | Sequence) -> Any:
+        """
+        Index single and multiple items from the dictionary.
+
+        Similar behavior to pandas.DataFrame.__getitem__.
+
+        Examples
+        --------
+        >>> components = ComponentsStore()
+        >>> components["generator"] = Generators()
+        >>> components["storage"] = Storage()
+        >>> components["generator"]
+        'Generator class instance'
+        >>> components[["generator", "storage"]]
+        >>> components[["generator", "storage"]]
+        ['Generator class instance', 'Storage class instance']
+        """
+        if isinstance(item, (list, set)):
+            return [self[key] for key in item]
+        else:
+            return super().__getitem__(item)
+
+    def __delattr__(self, name: str) -> None:
+        """
+        Is invoked when del object.member is called.
+        """
+        del self[name]
+
+    _re_pattern = re.compile("[a-zA-Z_][a-zA-Z0-9_]*")
+
+    def __dir__(self) -> list[str]:
+        """
+        Return a list of object attributes including dynamic ones from the dictionary keys.
+        """
+        dict_keys = [
+            k for k in self.keys() if isinstance(k, str) and self._re_pattern.match(k)
+        ]
+        obj_attrs = list(dir(super()))
+        return dict_keys + obj_attrs
 
-    def get_active_assets(
-        self,
-        investment_period: int | str | Sequence | None = None,
-    ) -> pd.Series:
+    def __iter__(self) -> Any:
         """
-        Get active components mask of componen type in investment period(s).
-
-        A component is considered active when:
-        - it's active attribute is True
-        - it's build year + lifetime is smaller than the investment period (if given)
-
-        Parameters
-        ----------
-        n : pypsa.Network
-            Network instance
-        c : str
-            Component name
-        investment_period : int, str, Sequence
-            Investment period(s) to check for active within build year and lifetime. If none
-            only the active attribute is considered and build year and lifetime are ignored.
-            If multiple periods are given the mask is True if the component is active in any
-            of the given periods.
-
-        Returns
-        -------
-        pd.Series
-            Boolean mask for active components
+        Make class iterable over the values.
+        #TODO: Do we want this? So list like behavior or dict like behavior?
```

Comment:
Dict-like might be better? But list-like is also ok, if component type can be inferred in another way.